rep.go

```diff
@@ -81,6 +93,242 @@ func (rep *repSocket) SetOption(name string, value interface{}) error {
 	return rep.sck.SetOption(name, value)
 }
 
+type repMsg struct {
+	conn *Conn
+	msg  Msg
+}
+
+type repReader struct {
+	ctx   context.Context
+	mu    sync.Mutex
+	conns []*Conn
+	msgC  chan repMsg
+	state *repState
+}
+
+func newRepReader(ctx context.Context, state *repState) *repReader {
+	const qsize = 10
+	return &repReader{
+		ctx:   ctx,
+		msgC:  make(chan repMsg, qsize),
+		state: state,
+	}
+}
+
+func (r *repReader) addConn(c *Conn) {
+	go r.listen(r.ctx, c)
+	r.mu.Lock()
+	r.conns = append(r.conns, c)
+	r.mu.Unlock()
+}
+
+func (r *repReader) rmConn(conn *Conn) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	cur := -1
+	for i := range r.conns {
+		if r.conns[i] == conn {
+			cur = i
+			break
+		}
+	}
+	if cur >= 0 {
+		r.conns = append(r.conns[:cur], r.conns[cur+1:]...)
+	}
+}
+
+func (r *repReader) read(ctx context.Context, msg *Msg) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case repMsg := <-r.msgC:
+		if repMsg.msg.err != nil {
+			return repMsg.msg.err
+		}
+		pre, innerMsg := splitReq(repMsg.msg)
+		if pre == nil {
+			return xerrors.Errorf("zmq4: invalid REP message")
+		}
+		*msg = innerMsg
+		r.state.Set(repMsg.conn, pre)
+	}
+	return nil
+}
+
+func (r *repReader) listen(ctx context.Context, conn *Conn) {
+	defer r.rmConn(conn)
+	defer conn.Close()
+
+	for {
+		msg := conn.read()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if msg.err != nil {
+				return
+			}
+			r.msgC <- repMsg{conn, msg}
+		}
+	}
+}
+
+func (r *repReader) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var err error
+	for _, conn := range r.conns {
+		e := conn.Close()
+		if e != nil && err == nil {
+			err = e
+		}
+	}
+	r.conns = nil
+	return err
+}
+
+func splitReq(envelope Msg) (preamble [][]byte, msg Msg) {
+	for i, frame := range envelope.Frames {
+		if len(frame) == 0 {
```

Comment:
```suggestion
		if len(frame) != 0 {
		    continue
		}
```

to reduce horizontal estate footprint :)